src/plugins/intel_npu/src/al/include/intel_npu/al/icompiler.hpp

```diff
@@ -22,48 +23,110 @@
+     * pairs.
+     */
+    std::optional<size_t> relatedDescriptorIndex;
+
+    /**
+     * @brief The friendly name of the node extracted from the IR model.
+     * @details In some cases, this field is required for constructing a dummy model which uses the same input/output
+     * metadata as the original IR model.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added by the
+     * compiler).
+     */
+    std::string nodeFriendlyName;
+
+    /**
+     * @brief The names of the output tensors extracted from the IR model.
+     * @details In some cases, this field is required for constructing a dummy model which uses the same input/output
+     * metadata as the original IR model.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added by the
+     * compiler).
+     */
     std::unordered_set<std::string> outputTensorNames;
-    ov::element::Type_t precision;
-    ov::PartialShape originalShape;
-    ov::PartialShape transposedShape;
-};
 
-/**
- * @brief A helper map to represent descriptions for inputs and outputs
- * of a network
- */
-using IONodeDescriptorMap = std::unordered_map<std::string, IONodeDescriptor>;
+    /**
+     * @brief The shape extracted from the IR model.
+     * @details The values may differ from the ones found in "shapeFromCompiler" if batching is to be handled by the
+     * plugin.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added
+     * by the compiler).
+     */
+    std::optional<ov::PartialShape> shapeFromIRModel = std::nullopt;
+};
 
 struct NetworkMetadata final {
     std::string name;
 
-    std::vector<std::string> inputNames;
-    std::vector<std::string> outputNames;
-    std::vector<std::string> stateNames;
-    std::vector<std::string> shapeNames;
+    std::vector<IODescriptor> inputs;
+    std::vector<IODescriptor> outputs;
+    std::vector<IODescriptor> profilingOutputs;
 
-    IONodeDescriptorMap parameters;
-    IONodeDescriptorMap results;
-    IONodeDescriptorMap states;
-    IONodeDescriptorMap shapes;
-    IONodeDescriptorMap profilingOutputs;
+    size_t numStreams = 1;
 
-    std::unordered_map<std::string, size_t> inputOrder;
-    std::unordered_map<std::string, size_t> outputOrder;
+    /**
+     * @brief Binds the (state input, state output) and (dynamic tensor, shape tensor) pairs using the
+     * "relatedDescriptorIndex" attribute.
+     * @details For state inputs, the "relatedDescriptorIndex" value is set to the index of the output which bears the
+     * same name. The reverse is also applied.
+     *
+     * For shape tensors, the lookup is performed in the same container (inputs or outputs). The value is once again set
+     * to the index of the entry which bears the same name.
+     */
+    void bindRelatedDescriptors();
```

Comment:
Seems this is only called by the compiler, so I suggest to move the implementation into the compiler.

Reply:
We'll continue [here](https://github.com/intel-innersource/applications.ai.vpu-accelerators.vpux-plugin/pull/10348#discussion_r1633422875), so I'll mark this one as resolved.